docs/architecture-decisions/flag-type.md

```diff
@@ -0,0 +1,144 @@
+---
+status: proposed
+author: @andreyturkov
+created: 2025-08-14
+updated: 2025-08-14
+---
+
+# Extending Flag Definition with a Type Property
+
+## Background
+
+Currently, `flagd` has inconsistent behavior in type validation between its `Resolve<T>` and `ResolveAll` API methods.
+The `Resolve<T>` method validates the evaluated flag variant against the type `T` requested by the client, while `ResolveAll` validates it against the type of the `defaultVariant` specified in the flag's definition.
+This discrepancy can lead to situations where a flag evaluation succeeds with one method but fails with the other, depending on the evaluation context and the variant returned.
+This inconsistent behavior is further detailed in bug report #1481.
+
+The root cause of this issue is the absence of a dedicated, authoritative type definition for the flag itself.
+Instead, the type is inferred from the `defaultVariant` or API itself (`T` from `Resolve<T>`), which is not always a reliable source of truth for all possible variants.
+This problem is getting worse by the planned support for code-defined defaults (as detailed in the [Support Code Default ADR](https://github.com/open-feature/flagd/blob/main/docs/architecture-decisions/support-code-default.md)), which allows the `defaultVariant` to be `null`.
+This makes it impossible to resolve the flag's type from the `defaultVariant`, increasing the risk of runtime errors.
+
+## Requirements
+
+* The new `flagType`field in the flag definition must be optional to ensure backward compatibility.
+* If the `flagType`field is present, `flagd` must validate that all variant values of the flag conform to this type during initialization.
```

Comment:
I _believe_ we could even validate this in the JSON schema with some backwards compatible changes to the definition: https://flagd.dev/schema/v0/flags.json